File: src/firebase/jsonTOcsv.py
import json
import csv
#import os
#import pandas as pd
#from datetime import datetime

# approach 2: just using json and csv library
# column mapping for the csv file
column_mapping = {
    "UTC Time": "timestamp.MatchUTC",
    "Unix Time Stamp": "timestamp.Match",
    "ID": "user.id",
    "ROLE": "user.admin",
    "GAME ID": "game.CurricularID",
    "GAME MODE": "game.mode",
    "DA Rep": "analytics.da_rep",
    "HINTS": "analytics.hints",
    "Hint Count": "analytics.hint_count",
    "Latin Square Order": "settings.latin_square_order",
    "HINT ORDER": "settings.hint_order",
    "CONJ": "session.conj",
    "ETSS": "session.etss",
    "ETSLO": "session.etslo",
    "Event Type": "event.type",
    "TF Given Answer": "response.tf_given_answer",
    "Correct": "response.correct",
    "MC Given Answer": "response.mc_given_answer",
    "MC Correct": "response.mc_correct"
}

# approach 3: using json and csv library with helper functions
import json
import csv
from datetime import datetime

# Flatten nested data for poses
def extract_poses(conjecture_data, game_id, user_id, role, timestamp):
    rows = []
    for pose, details in conjecture_data.items():
        row = {
            "UTC Time": details.get("StartUTC"),
            "Unix Time Stamp": details.get("Start"),
            "ID": user_id,
            "ROLE": role,
            "GAME ID": game_id,
            "GAME MODE": "default_mode",  # Adjust as needed
            "DA Rep": "null",  # Adjust if present
            "HINTS": "null",  # Adjust if present
            "Pose": pose,
            "Start Match": details.get("MatchUTC"),
        }
        rows.append(row)
    return rows

# Parse JSON and write to CSV
def map_and_convert_to_csv(json_data, output_file):
    rows = []

    # Iterate through games
    for game_name, game_details in json_data.items():
        curricular_id = game_details.get("CurricularID")
        for date, date_details in game_details.items():
            if isinstance(date_details, dict):
                for role, role_details in date_details.items():
                    user_id = role_details.get("UserId")
                    for timestamp, session_details in role_details.items():
                        if isinstance(session_details, dict):
                            conjecture_data = session_details.get("ConjectureId", {})
                            rows.extend(extract_poses(
                                conjecture_data, curricular_id, user_id, role, timestamp
                            ))

    # Write to CSV
    with open(output_file, mode='w', newline='', encoding='utf-8') as file:
        writer = csv.DictWriter(file, fieldnames=[
            "UTC Time", "Unix Time Stamp", "ID", "ROLE", "GAME ID", "GAME MODE", "DA Rep", "HINTS", "Pose", "Start Match"
        ])
        writer.writeheader()
        writer.writerows(rows)

# Main function
def main():
    # Load JSON
    input_file = "C:/Users/adamc/Downloads/exported-json-data-2024-11-03T21-19-02.293Z.json"
    output_file = "formatted_output.csv"

    with open(input_file, "r") as file:
        json_data = json.load(file)

    # Convert and write to CSV
    map_and_convert_to_csv(json_data, output_file)

if __name__ == "__main__":
    main()


# ===========================================================================================================================================
# approach 2: just using json and csv library 
# helper function to get nested value from a dictionary
# def get_nested_value(data, path):
#     keys = path.split('.')
#     for key in keys:
#         if isinstance(data, dict) and key in data:
#             data = data[key]
#         else:
#             return None
#     return data

# # Function to read json file
# def read_json(filename):
#     # Check if file exists
#     try:
#         with open(filename, 'r') as f:
#             data = json.load(f)
#     except:
#         raise Exception("Error reading json file")
#     return data


# # map flatten json to csv
# def map_to_csv_format(data):
#     # create a dictionary to store the csv data
#     mapped_data = {col: get_nested_value(data, path) for col, path in column_mapping.items()}
#     print(mapped_data)
#     return mapped_data

# # function to normalize json data
# # def normalize_json(data):
# #     new_data = dict()
# #     for key, value in data.items():
# #         if not isinstance(value, dict):
# #             new_data[key] = value
# #         else:
# #             for k, v in value.items():
# #                 new_data[key + "_" + k] = v

# #     return new_data                              
   
# # Function to generate csv file
# def generate_csv(mapped_data_list, output_fileName):
#     # extract headers from the first mapped data entry
#     headers = column_mapping.keys()

#     # write to csv file
#     with open(output_fileName, 'w', newline='') as f:
#         writer = csv.DictWriter(f, fieldnames=headers)
#         writer.writeheader() # write header row
#         for row in mapped_data_list:
#             writer.writerow(row) # write data rows

# # Function to write data to file
# # def write_to_file(data, filepath):
    
# #     try:
# #         with open(filepath, 'w') as f:
# #             f.write(data)
# #     except:
# #         raise Exception("Error writing to file")
    

# # main function
# def main(input_file, output_file):
#     # read the json file as python dictionary
#     # make this as a input from the user
#     data = read_json(filename= input_file)

    
#     mapped_data_list = [map_to_csv_format(entry) for entry in data.values()]

#     # flatten the json data
#     # flat_json = flatten_json(data)
#     # print("Flattened json: ", flat_json)

#     #save the csv data to a file
#     generate_csv(mapped_data_list, output_file)

#     # generate desired csv data 
#     # csv_data = generate_csv(data=flat_json)

#     # normalize the nested python dictionary
#     # new_data = normalize_json(data=data)
 
#     # pretty print the normalized data
#     # print("New dictionary: ", new_data)

# if __name__ == "__main__":
#     main(input_file='C:/Users/adamc/Downloads/exported-json-data-2024-11-03T21-19-02.293Z.json', output_file='data.csv')

   
# ===========================================================================================================================================
# apporoach 1: using pandas ==== library issues

# # Function to read json file
# def read_json(filename):
#     # Check if file exists
#     try:
#         with open(filename, 'r') as f:
#             data = json.load(f)
#     # If file does not exist, raise exception
#     except:
#         raise Exception("Error reading json file")
#     # Return data
#     return data

# def create_dataframe(data):
#     # Create dataframe
#     df = pd.DataFrame()
#     # looping throug each record
#     for d in data:
#         # normalize column levels 
#         record = pd.json_normalize(d)

#         # append it to the dataframe
#         df = df.append(record, ignore_index=True)   
#     return df

# # main function
# def main():
#     # read json file
#     data = read_json(filename= 'C:/Users/adamc/Downloads/exported-json-data-2024-10-30T14-15-57.374Z.json')
#     # generate the datarame for the array items in key
#     # key is the array of the json in which you are pulling from 
#     df = create_dataframe(data = data)


#     # rename the colums in dataframe
#     print("normalized columns", df.columns.to_list())
#     df.rename(columns = {'CurricularID':'GameID'}, inplace=True)

#     print("renamed columns", df.columns.to_list())

#     # write the dataframe to csv
#     df.to_csv('data.csv', index=False)

File: src/firebase/jsonTOcsv.js


export async function convertJsonToCsv(jsonFile, formattedGame, formattedStart, formattedEnd) {
    try {
        // Read JSON file content
        //const fileContent = await jsonFile.text();
        const jsonData = JSON.parse(jsonFile);
        
        // Headers for CSV
        const headers = [
            "UTC Time",
            "Unix Time Stamp",
            "ID",
            "ROLE",
            "GAME ID", 
            "GAME MODE",
            "DA Rep",
            "HINTS",
            "Hint Count",
            "Conj",
            "ETSS",
            "ETSLO",
            "Event Type",
            "TF Given Answer",
            "TF Correct",
            "MC Given Answer",
            "MC Correct",
            "Pose",
            "Start Match"
        ];

        // Process data
        const rows = [];
        // process nested structure
        for (const [date, dateData] of Object.entries(jsonData)) {
            for (const [role, roleData] of Object.entries(dateData)) {
                const userId = roleData.UserId ?? "null";
                
                for (const [timestamp, sessionData] of Object.entries(roleData)) {
                    if (timestamp === "UserId") continue; // Skip UserId entry
                    
                    // Extract game-level data
                    const daRep = sessionData.DaRep ?? "null";
                    const hintCount = sessionData.Hints?.HintCount ?? "null";
                    const latinSquareOrder = sessionData.LatinSquareOrder ?? "null";
                    
                    // Process each pose session
                    for (const [sessionId, sessionDetails] of Object.entries(sessionData)) {
                        if (sessionId === "DaRep" || sessionId === "GameStart" || 
                            sessionId === "GameStartGMT" || sessionId === "Hints" || 
                            sessionId === "LatinSquareOrder") continue;
                            
                        // Process poses
                        for (const [poseName, poseDetails] of Object.entries(sessionDetails)) {
                            if (poseName.startsWith("Pose") || poseName === "Intuition") {
                                rows.push({
                                    "UTC Time": poseDetails.StartGMT ?? "null",
                                    "Unix Time Stamp": poseDetails.Start ?? "null",
                                    "ID": userId,
                                    "ROLE": role,
                                    "GAME ID": sessionId,
                                    "GAME MODE": "default_mode",
                                    "DA Rep": daRep,
                                    "HINTS": "null",
                                    "Hint Count": hintCount,
                                    "Conj": "null",
                                    "ETSS": "null",
                                    "ETSLO": latinSquareOrder,
                                    "Event Type": "null",
                                    "TF Given Answer": "null",
                                    "TF Correct": "null",
                                    "MC Given Answer": "null",
                                    "MC Correct": "null",
                                    "Pose": poseName,
                                    "Start Match": poseDetails.MatchGMT ?? "null"
                                });
                            }
                        }
                    }
                }
            }
        }

        // Create CSV content
        const csvContent = [
            headers.join(','),
            ...rows.map(row => headers.map(header => `"${row[header] ?? "null"}"`).join(','))
        ].join('\n');

        // Create and download file
        const blob = new Blob([csvContent], { type: 'text/csv' });
        const url = URL.createObjectURL(blob);
        const link = document.createElement('a');
        link.href = url;
        link.download = `${formattedgame}_event_log_${formattedStart}_to_${formattedEnd}.csv`;
        link.click();
        URL.revokeObjectURL(url);

        return { success: true, rowCount: rows.length, csvContent: csvContent };
    } catch (error) {
        console.error('Conversion failed:', error);
        throw new Error(`Failed to convert file: ${error.message}`);
    }
}

// function extractPoses(conjectureData, gameId, userId, role, timestamp) {
//     return Object.entries(conjectureData).map(([pose, details]) => ({
//         "UTC Time": details.StartUTC ?? "null",
//         "Unix Time Stamp": details.Start ?? "null",
//         "ID": userId,
//         "ROLE": role,
//         "GAME ID": gameId,
//         "GAME MODE": "default_mode",
//         "DA Rep": "null",
//         "HINTS": "null",
//         "Hint Count": "null",
//         "Conj": "null",
//         "ETSS": "null",
//         "ETSLO": "null",
//         "Event Type": "null",
//         "TF Given Answer": "null",
//         "TF Correct": "null",
//         "MC Given Answer": "null",
//         "MC Correct": "null",
//         "Pose": pose,
//         "Start Match": details.MatchUTC ?? "null"
//     }));
// }



// const fs = require("fs");
// const os = require("os");
// const path = require("path");

// // Function to extract poses from conjecture data
// export function extractPoses(conjectureData, gameId, userId, role, timestamp) {
//     const rows = [];
//     for (const [pose, details] of Object.entries(conjectureData)) {
//         rows.push({
//             "UTC Time": details.StartUTC || "null",
//             "Unix Time Stamp": details.Start || "null",
//             "ID": userId || "null",
//             "ROLE": role || "null",
//             "GAME ID": gameId || "null",
//             "GAME MODE": "default_mode",
//             "DA Rep": "null",
//             "HINTS": "null",
//             "Pose": pose || "null",
//             "Start Match": details.MatchUTC || "null",
//         });
//     }
//     return rows;
// }
// // Function to map JSON data and convert to CSV
// export function mapAndConvertToCSV(jsonData) {
//     const rows = [];
//     const headers = [
//         "UTC Time",
//         "Unix Time Stamp",
//         "ID",
//         "ROLE",
//         "GAME ID",
//         "GAME MODE",
//         "DA Rep",
//         "HINTS",
//         "Pose",
//         "Start Match",
//     ];
//     // Extract poses from JSON data and add to rows
//     for (const [gameName, gameDetails] of Object.entries(jsonData)) {
//         const curricularId = gameDetails.CurricularID || "null";
//         for (const [date, dateDetails] of Object.entries(gameDetails)) {
//             if (typeof dateDetails === "object") {
//                 for (const [role, roleDetails] of Object.entries(dateDetails)) {
//                     const userId = roleDetails.UserId || "null";
//                     for (const [timestamp, sessionDetails] of Object.entries(
//                         roleDetails
//                     )) {
//                         if (typeof sessionDetails === "object") {
//                             const conjectureData =
//                                 sessionDetails.ConjectureId || {};
//                             rows.push(
//                                 ...extractPoses(
//                                     conjectureData,
//                                     curricularId,
//                                     userId,
//                                     role,
//                                     timestamp
//                                 )
//                             );
//                         }
//                     }
//                 }
//             }
//         }
//     }

//     // Create CSV content from rows
//     const csvContent =
//         headers.join(",") +
//         "\n" +
//         rows
//             .map((row) =>
//                 headers.map((header) => `"${row[header] || "null"}"`).join(",")
//             )
//             .join("\n");

//     // Save to Downloads folder
//     const downloadsPath = path.join(os.homedir(), "Downloads");
//     const outputPath = path.join(downloadsPath, "formatted_output.csv");

//     fs.writeFileSync(outputPath, csvContent);
//     console.log(`CSV file saved to ${outputPath}`);
// }

// // Main function to load JSON file and convert to CSV
// export function main() {
//     const inputFilePath = "C:/Users/adamc/Downloads/exported-json-data-2024-11-03T21-19-02.293Z.json"; // Update to your local JSON file path

//     // Load JSON file
//     const jsonData = JSON.parse(fs.readFileSync(inputFilePath, "utf-8"));

//     // Convert and save CSV
//     mapAndConvertToCSV(jsonData);
// }

// main();
